lib/facter/util/resolvers/http.rb

```diff
@@ -10,34 +10,39 @@ class << self
           CONNECTION_TIMEOUT = 0.6
           SESSION_TIMEOUT = 5
 
-          # Makes a GET http request and returns its response.
+          # Makes a GET HTTP request and returns its response.
           #
-          # Params:
-          # url: String which contains the address to which the request will be made
-          # headers: Hash which contains the headers you need to add to your request.
-          #          Default headers is an empty hash
-          #          Example: { "Accept": 'application/json' }
-          # timeouts: Hash that includes the values for the session and connection timeouts.
-          #          Example: { session: 2.4. connection: 5 }
-          #
-          # Return value:
-          # is a string with the response body if the response code is 200.
-          # If the response code is not 200, an empty string is returned.
-          def get_request(url, headers = {}, timeouts = {})
-            make_request(url, headers, timeouts, 'GET')
+          # @param url [String] the address to which the request will be made.
+          # @param headers [Hash] the headers you need to add to your request.
+          #   Defaults to an empty hash.
+          # @param timesouts [Hash] Values for the session and connection
+          #   timeouts.
+          # @param http_port [Integer] The port number used to make the
+          #   request. Defaults to 80.
+          # @param proxy_addr [String] The address of a proxy host.
+          # @returns [String] the response body if the response code is 200.
+          #   If the response code is not 200, an empty string is returned.
+          # @example
+          #   get_request(https://example.com, { "Accept": 'application/json' }, { session: 2.4. connection: 5 })
+          def get_request(url, headers = {}, timeouts = {}, http_port: 80, proxy_addr: ENV['http_proxy'])
```

Comment:
Previously, if the `http_proxy'` environment variable was set, then the call to `Net::HTTP.new(parsed_url.host)` might or might not use the proxy, depending on whether we're connecting to the loopback:

https://github.com/ruby/ruby/blob/1f4d4558484b370999954f3ede7e3aa3a3a01ef3/lib/uri/generic.rb#L1533

or the target host matches the `no_proxy` environment variable:

https://github.com/ruby/ruby/blob/1f4d4558484b370999954f3ede7e3aa3a3a01ef3/lib/uri/generic.rb#L1541

If we always pass `ENV['http_proxy']` through to the `Net::HTTP.new` method, then I think(?) it will always use that explicit proxy.

Maybe define a `DEFAULT_HOST` object and default to the proxy_addr option to that:

```
DEFAULT_HOST = Object.new.freeze
def get_request(url, headers = {}, timeouts = {}, http_port: 80, proxy_addr: DEFAULT_HOST)
```

Then check if the option has been overridden (to possibly `nil`) like:

```ruby
if proxy_addr == DEFAULT_HOST
  Net::HTTP.new(host, port)
else
  Net::HTTP.new(host, port, proxy_addr, proxy_port)
```

Then have the cloud resolvers explicitly pass `proxy_addr: nil`